Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/checkins/cryptography/CheckInCryptography.kt

```diff
@@ -1,25 +1,135 @@
 package de.rki.coronawarnapp.presencetracing.checkins.cryptography
 
+import androidx.annotation.VisibleForTesting
+import de.rki.coronawarnapp.covidcertificate.common.cryptography.AesCryptography
 import de.rki.coronawarnapp.presencetracing.checkins.CheckIn
+import de.rki.coronawarnapp.server.protocols.internal.pt.CheckInOuterClass.CheckInRecord
 import de.rki.coronawarnapp.server.protocols.internal.pt.CheckInOuterClass.CheckInProtectedReport
-import de.rki.coronawarnapp.server.protocols.internal.pt.TraceWarning
+import de.rki.coronawarnapp.util.HashExtensions.toSHA256
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.seconds
+import de.rki.coronawarnapp.util.encoding.base64
 import de.rki.coronawarnapp.util.security.RandomStrong
+import de.rki.coronawarnapp.util.toProtoByteString
+import okio.ByteString
+import okio.ByteString.Companion.decodeHex
+import okio.ByteString.Companion.toByteString
+import org.joda.time.Instant
+import javax.crypto.Mac
+import javax.crypto.spec.IvParameterSpec
+import javax.crypto.spec.SecretKeySpec
 import javax.inject.Inject
 import kotlin.random.Random
 
 class CheckInCryptography @Inject constructor(
-    @RandomStrong private val randomSource: Random,
+    @RandomStrong private val secureRandom: Random,
+    private val aesCryptography: AesCryptography
 ) {
     fun encrypt(
         checkIn: CheckIn,
         transmissionRiskLevel: Int
     ): CheckInProtectedReport {
-        // TODO
-        throw NotImplementedError()
+
+        val checkInRecord = createCheckInRecord(checkIn.checkInStart, checkIn.checkInEnd, transmissionRiskLevel)
+        val encryptionKey = getEncryptionKey(checkIn.traceLocationId.toByteArray())
+        val iv = getCryptographicSeed()
+        val encryptedCheckInRecord =
+            aesCryptography.encrypt(encryptionKey.toByteArray(), checkInRecord.toByteArray(), IvParameterSpec(iv))
+        val macKey = getMacKey(checkIn.traceLocationId.toByteArray()).toByteArray()
+        val mac = getMac(macKey, iv, encryptedCheckInRecord)
+
+        return CheckInProtectedReport.newBuilder()
+            .setEncryptedCheckInRecord(encryptedCheckInRecord.toByteString().toProtoByteString())
+            .setIv(iv.toByteString().toProtoByteString())
+            .setLocationIdHash(checkIn.traceLocationIdHash.toProtoByteString())
+            .setMac(mac.toByteString().toProtoByteString())
+            .build()
+    }
+
+    fun decrypt(
+        checkInProtectedReport: CheckInProtectedReport,
+        locationId: ByteArray
```

Comment:
decryption is happening during the matching process, where we have `TraceWarningPackage`s that has either 
`TraceTimeIntervalWarning` or `CheckInProtectedReport` and both have `locationIdHash` only. how locationId is supposed to come from ?

Reply:
In the database for check-ins, you should have Location ID and Location ID Hash. You look up the record by Location ID Hash and then obtain the Location ID.  See TechSpec, _Deriving TraceTimeIntervalWarnings from CheckInProtectedReports_, bullet 2 and 3.

Reply:
@jurajkusnier  can you please  use  `traceLocationId: TraceLocationId` instead of `locationId: ByteArray` to follow the same syntax for locationId as everywhere else ?